plugins/amqp/src/plugin.cpp

```diff
@@ -0,0 +1,748 @@
+    if (auto* sasl_method = get_if<std::string>(&settings, "sasl_method")) {
+      if (*sasl_method == "plain")
+        config.sasl_method = AMQP_SASL_METHOD_PLAIN;
+      else if (*sasl_method == "external")
+        config.sasl_method = AMQP_SASL_METHOD_EXTERNAL;
+      else
+        return caf::make_error(ec::parse_error,
+                               fmt::format("invalid SASL method: {}",
+                                           *sasl_method));
+    }
+    return amqp_engine{std::move(config)};
+  }
+
+  /// Constructs an AMQP engine from a typed configuration.
+  /// @param config The AMQP configuration.
+  explicit amqp_engine(amqp_config config)
+    : config_{std::move(config)}, conn_{amqp_new_connection()} {
+    TENZIR_ASSERT(conn_ != nullptr);
+    TENZIR_DEBUG("constructing AMQP engine with the following parameters:");
+    TENZIR_DEBUG("- hostname: {}", config_.hostname);
+    TENZIR_DEBUG("- port: {}", config_.port);
+    TENZIR_DEBUG("- ssl: {}", config_.ssl);
+    TENZIR_DEBUG("- vhost: {}", config_.vhost);
+    TENZIR_DEBUG("- max_channels: {}", config_.max_channels);
+    TENZIR_DEBUG("- frame_size: {}", config_.frame_size);
+    TENZIR_DEBUG("- heartbeat: {}", config_.heartbeat);
+    TENZIR_DEBUG("- username: {}", config_.username);
+    TENZIR_DEBUG("- password: ***");
+    TENZIR_DEBUG("- SASL method: {}", static_cast<int>(config_.sasl_method));
+    TENZIR_DEBUG("creating new TCP socket");
+    if (config_.ssl)
+      socket_ = amqp_ssl_socket_new(conn_);
+    else
+      socket_ = amqp_tcp_socket_new(conn_);
+    TENZIR_ASSERT(socket_ != nullptr);
+  }
+
+  ~amqp_engine() {
+    if (not conn_)
+      return;
+    TENZIR_DEBUG("closing AMQP connection");
+    auto reply = amqp_connection_close(conn_, AMQP_REPLY_SUCCESS);
+    if (auto err = to_error(reply))
+      TENZIR_DEBUG(err);
+    TENZIR_DEBUG("destroying AMQP connection");
+    auto status = amqp_destroy_connection(conn_);
+    if (auto err = to_error(status, "failed to destroy AMQP connection"))
+      TENZIR_WARN(err);
+  }
+
+  // Be a move-only handle type.
+  amqp_engine(amqp_engine&) = delete;
+  auto operator=(amqp_engine&) -> amqp_engine& = delete;
+  auto operator=(amqp_engine&&) -> amqp_engine& = delete;
+
+  amqp_engine(amqp_engine&& other) noexcept
+    : config_{std::move(other.config_)},
+      conn_{other.conn_},
+      socket_{other.socket_},
+      channel_{other.channel_} {
+    other.conn_ = nullptr;
+    other.socket_ = nullptr;
+    other.channel_ = 0;
+  }
+
+  /// Connects to the server by opening a socket and logging in.
+  auto connect() -> caf::error {
+    if (auto err = open_socket())
+      return err;
+    return login();
+  }
+
+  /// Opens a channel.
+  /// @param number The channel number.
+  auto open(amqp_channel_t number) -> caf::error {
+    TENZIR_DEBUG("opening AMQP channel {}", number);
+    amqp_channel_open(conn_, number);
+    return to_error(amqp_get_rpc_reply(conn_));
+  }
```

Comment:
Why expose internals if we don't have to? Unless the user would do stuff with the RPC reply, I don't think it makes sense to expose it. Arguing with KISS, this would be simpler and therefore more desirable

Reply:
Resolving this as "not actionable" in the grand scheme of things.

(This feedback is still appreciated, so please keep it coming!)